---
title: "Ontology and Corpus Aided Searching of Complex Data Resources"
author: "RoGe, VC, RaGo"
date: "2023-06-23"
output: 
  bookdown::html_document2: 
    number_sections: false
vignette: >
  %\VignetteIndexEntry{Ontology and Corpus Aided Searching of Complex Data Resources}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

bibliography: references.bib

link-citations: true

format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("gwasCatSearch")
library("kableExtra")
library("DT")
```

## Introduction


Larger and more complex biomedical data sets are becoming increasingly available and they provide substantial opportunities for integration across disparate data source to enable the discovery of medically relevant relationships. Realizing the potential for discovery is hampered by many challenges including the complexity of the data and the inherent challenge of working with phenotypes and experimental factors described primarily in natural language. To work with these textual descriptions there are a plethora of ontologies and vocabularies that have been proposed and which are in use such as Unified Medical Language System (UMLS, @UMLS2004) , Medical Subject Headings (MeSH), Human Phenotype Ontology (HPO, @HPO2024), and the Experimental Factor Ontology (EFO, @maloneModelingSampleVariables2010). And a similar wide range of objectives such as finding, and matching, relevant phenotypes in large data resources such as NHANES or Biobanks.

We explore some of these challenges in the context of the GWAS catalog, (@sollisNHGRIEBIGWASCatalog2023) which reports on a set of published curated GWAS association studies where the phenotypic traits reported in published articles are mapped to terms in the EFO. GWAS studies report on assocations between single nucleotide polymorphisms (SNPs), which are variants in the human genome, and specific traits. To facilitate searching for traits we use a search strategy that combines the use of a dictionary (corpus based) search combined with an ontology based search.  We create a dictionary search capability that is built on the textual descriptions from the GWAS catalog and EFO which we combine with a set of tools designed to make use of the relationships and concepts in the ontology.  We demonstrate how these tools, when used together, enable searching the GWAS catalog. The approach is similar to that of @phenotagger where they describe a method that combines both dictionary and machine learning-based methods to identify (HPO) concepts in unstructured biomedical text.

In particular we examine the use of corpus search tools, as exemplified in the `corpustools` package
@atteveldtCorpustoolsManagingQuerying2023, to help search through the set of traits that are
reported by the GWAS catalog. Since these investigator defined traits have been hand curated by the
EBI and mapped to terms in the  we can also examine how the structure that mapping provides can be used to aid the analysis and interpretation of the data.

Among the curation activities is a process of mapping investigator provided traits to
the Experimental Factor Ontology (EFO) @maloneModelingSampleVariables2010. 
In part this allows one to search for
studies about a particular disease, but for a variety of reasons this
search can be quite complicated. Here we report on tools and
approaches that can be used to help find GWAS studies according to
specific criteria. We create two searchable corpora, one that is EFO
centric and the other that is GWAS trait centric. There are a number of 
anomalies in both the EFO and the GWAS catalog that make it essential that
we have effective tools for cross-referencing searches. Our software is contained in the `gwasCatSearch` package (@gentlemanGwasCatSearchPackageSupport2024a) 
which is open source and freely available.

## Materials and Methods

### Data

Since 2010, delivery and development of the Catalog
has been a collaborative project between the EMBL-EBI and NHGRI. Full details of
the data and the date it was obtained are given in an appendix. As of
2024-01-30, the GWAS Catalog contains 6715 publications, 571148 top
associations and 67322 full summary statistics. GWAS Catalog data is
currently mapped to Genome Assembly GRCh38.p14 and dbSNP Build 156.
The curation process used is described on their web page, 
<https://www.ebi.ac.uk/gwas/docs/methods/curation>. Of particular
interest is their comment:
*Each study is also assigned a trait that best represents the phenotype under investigation. When multiple traits are analysed in the same study either multiple entries are created, or individual SNPs are annotated with their specific traits.*. Which seems to mean that in some situations individual SNPs will be annotated with a trait, while in other cases collections of SNPs will be associated with a single trait.

We obtain data from the GWAS Catalog and transform it into a set of tables using tools in the GWASCatalogSearchDB (https://github.com/ccb-hms/GWASCatalogSearchDB).
Our processing produces a number of tables that can be accessed via a SQL database; we use SQLite (@sqlite2020hipp). These tables include the raw data from the EBI, several tables organized around the Experimental Factor Ontology (EFO), the Uber-anatomy ontology (UBERON) @Uberon2012.

### Ontologies

An ontology is a set of concepts and categories that describe a subject area or domain of knowledge that expresses their properties and the relations between them. Ontologies form one of the bases for knowledge engineering and representation and are important in understanding and referencing biomedical entities. One of the challenges that needs to be addressed is the fact that the same concept might usefully appear in different ontologies and users will need tools to help identify similar concepts across ontologies. A second challenge is that to be useful the ontologies must be accessible and typically that means using the web or a database implementation. There are two constructs that help with this, the compact uniform resource identifiers (CURIEs), and the internationalized resource identifiers (IRIs). A CURIE has three parts: a prefix; a delimiter; and a unique local (to the ontology described by the prefix) identifier from the given nomenclature. The CURIE EFO:0005305 which represents the concept *atrioventricular node disease* has a prefix of EFO, a delimiter of a colon (:) and a unique local identifier of 0005305.  The corresponding IRI for that term is http://www.ebi.ac.uk/efo/EFO_0005305, which gives a web location that can be examined by a user. Note that the delimiter in the IRI is an underscore, rather than a colon. We will use both in our discussion. The terms in the EFO ontology
are related to each other via a directed acyclic graph. By construction there is a single root node. More specific terms, such as the relationship between cancer and pancreatic cancer are encoded with pancreatic cancer being the **child** of cancer, while cancer is the **parent** of pancreatic cancer.  The descriptions can become more refined, and we define the **descendents** of a node to be the set of child nodes, where one recurses down to the point where a node has no children.  Such a node is referred to as a *leaf*.  Similarly **ancesters** are the parents, again applied recursively up to the root node of the ontology.  Our software contains tools to obtain the children, descendents, parents, and ancestors as well as to perform other operations on the ontology.  This organization yields substantial benefits, as all diseases will be organized under the node labeled *Disease* which has CURIE EFO:0000408. 

Disease location information was extracted directly from EFO statements of the form:  X has_disease_location Y, where Y is typically an UBERON term representing an anatomical location. If a term does not have an explicitly stated location relationship, we determine if it has an inferred location. This is done by recursively checking if a parent in the ontology hierarchy has a location. The first location found is used, and if none is found then it is marked as missing. 
For example, *bronchitis*  (EFO:0009661) does not have an explicitly stated location in EFO, but it inherits one from its immediate parent (*bronchial disease*), which has location *bronchus*, (UBERON:0002185).  Of the `r NROW(efo_df)` EFO terms `r sum(!is.na(efo_df$DiseaseLocation))` terms have a disease location.

### Aligning EFO and the GWAS Catalog

Our software contains a dataframe in which there is one row for each unique combination of GWAS catalog study ID and EFO term. The GWAS catalog dataframe columns are labeled `STUDY.ACCESSION`: the GWAS catalog study identifier; `DISEASE.TRAIT`: the GWAS catalog text description of the trait; `MAPPED_TRAIT`: the textual description of the trait from EFO that this study was *directly* mapped to; `MAPPED_TRAIT_URI`: the URI for the mapped trait; `MAPPED_TRAIT_CURIE`: the CURIE for the mapped trait.

We also provide a dataframe representing the EFO terms, with one row for each term in the EFO ontology and columns labeled `Subject`: the corresponding EFO CURIE; `Object`: the textual description of the trait; `IRI`: the IRI for that term; `DiseaseLocation`: the UBERON CURIE for the location of the `trait` if there is one; `Direct`: the number of GWAS catalog studies that were mapped directly to that trait; `Inherited`: the number of traits that were mapped to terms that are more specific (according to the ontology) than the `Object`.

These two dataframes are summarized in Tables \@ref(tab:tab1a) and \@ref(tab:tab1b). For the EFO data there are two columns that were not included in the summary table, `Direct` which contains the count of GWAS studies that were directly mapped to that trait and `Inherited` which is the count of GWAS studies that are inherited from children of the EFO term.


```{r tab1a, echo=FALSE}
ttz = data.frame(t(t(sapply(gwc_df,function(x)length(unique(x))))))
ttz = cbind(rownames(ttz), ttz)
rownames(ttz) = NULL
names(ttz) = c("Field", "Unique values")
kable_styling(kbl(ttz, caption="Overview of GWAS Catalog Data"), bootstrap_options = "striped", full_width = FALSE)
```


```{r tab1b, echo=FALSE}
ttz = data.frame(t(t(sapply(efo_df,function(x)length(unique(x))))))[1:4,]
rn = c("Subject: EFO CURIE", "Object: Trait Name", "IRI: IRI", "DiseaseLocation: UBERON CURIE")
ttz = cbind(rn, ttz)
rownames(ttz) = NULL
colnames(ttz) = c("Field", "Unique values")
kable_styling(kbl(ttz, caption="Overview of EFO Data"), bootstrap_options = "striped", full_width = FALSE)
```

```{r tab1c,echo=FALSE}
ee = head(efo_df[,-c(3,4)][rev(order(efo_df[,"Direct"])),],8)
rownames(ee) = NULL
kable_styling(kbl(ee, caption="The top 8 records in efo_df, ordered by frequency of direct annotation to GWAS."), bootstrap_options = "striped", full_width = FALSE)
```


In some sense studies that map to leaf nodes in the ontology are describing quite specific concepts, while studies that map to non-leaf nodes are likely to be concepts that are more general.
FIXME: we need an image of some part of the ontology to try to make this clear to the reader

Tools for representing and organizing biomedical entities and concepts are key components of
any system that process disease, medical or epidemiological data.  There are many challenges among them is the need to have some way to have two different systems align on some concepts. On the seminatic web Uniform Resource Identifiers (URIs) are used to reference named entities. So URI basically give a web address where more details on the concept can be found.

```{r tryfig, echo=FALSE, eval=FALSE, message=FALSE, fig.cap="Terms and relations in EFO"}
if (!requireNamespace("ontoProc")) stop("install ontoProc with BiocManager to run this chunk")
mytags = c("EFO:0004747", "EFO:0004346", "EFO:0004529", "EFO:0006801", 
"MONDO:0004975", "HP:0003124", "EFO:0005210", "EFO:0007849")
suppressMessages({ef = ontoProc::getOnto("efoOnto")})
onto_plot2(ef, mytags)
```

Given the structure of the ontology we can introduce the concept of an *inherited* mapping. A mapping at a node E is inherited if it was a direct mapping to some descendant of E.  Basically this is
providing a way to find, and hence examine, the EFO terms that map to more specific versions of an EFO term.  So, while we might want to know the GWAS catalog traits that map to cancer, we may also want to find the traits that map to more specific types of cancer.  This concept allows us to do that in a rigorous way.

```{r ontoHelperFuns, echo=FALSE}

##a leaf is a node that has no children
##the root is a node that has no parents

isLeaf = function(x) {
  descend = EFOdescendants(x)
  sp = split(descend, names(descend))
  ans = sapply(sp, function(x) length(x) < 2)
  ans
}

t1 = isLeaf(c('MONDO:0019640','EFO:0004194'))

z2 = EFOdescendants("EFO:0005741")
z3 = EFOancestors("EFO:0005741")

isRoot = function(x) {
  parents=EFOancestors(x)
  sp = split(parents, names(parents))
  ans = sapply( sp, function(x) length(x) < 2)
  ans
}
```


```{r cmpTableofMappingCounts, echo=FALSE }
##we are looking at how many rows in gwc_df each Study accession number
##has - so each of these corresponds to a unique EFO term
pp = table(gwc_df$STUDY.ACCESSION)
ptab = table(pp)
ptab2 = c(ptab[1:9], sum(ptab[10:length(ptab)]))
names(ptab2)[10] = "10+"
kable_styling(kbl(t(ptab2), caption="GWAS Catalog Studies by number of EFO terms per Study"), bootstrap_options = "striped", full_width = FALSE)
```


There are `r length(unique(gwc_df$STUDY.ACCESSION))` studies reported.  Each study can be mapped to one or more EFO terms. We counted the number of EFO terms each study was mapped to and report those in Table \@ref(tab:cmpTableofMappingCounts). The counts range from 1 to `r max(pp)` with the study `r names(pp)[which.max(pp)]` having the most and the vast majority are mapped to only one EFO term.

```{r studiesperPMID, echo=FALSE}
xx = DBI::dbGetQuery(gwasCatSearch_dbconn(), "SELECT PUBMEDID, \"STUDY.ACCESSION\" FROM gwascatalog_metadata")
tt = sort(table(xx$PUBMEDID),dec=T)
```

```{r checkTraitsBackman, echo=FALSE, warning=FALSE }
#no duplicates in saB, 175 in dupT and 7K in the CURIEs
saB = xx$STUDY.ACCESSION[xx$PUBMEDID == names(tt)[1]]
gwcB = gwc_df[gwc_df$STUDY.ACCESSION %in% saB, ]
dupT = sum(duplicated(gwcB$DISEASE.TRAIT))
dupC = sum(duplicated(gwcB$MAPPED_TRAIT_CURIE))
```

The GWAS catalog also provides metadata in the form of a table with one entry per STUDY.ACCESSION number and other variables such as the PUBMEDID for the associated publication, more complete information on these tables is in the Appendix. A single paper can give rise to more than one GWAS catalog study. We find that there are `r sum(tt > 100)` publications which have more than 100 GWAS studies associated with them and the paper associated with PMID `r names(tt)[1]` @Backman2021 has `r tt[1]` studies associated with it.  In the Backman *et al* paper the authors report ``We identified 12 million coding variants, including around 1 million loss-of-function and around 1.8 million deleterious missense variants. When these were tested for association with 3,994 health-related traits, we found 564 genes with trait associations at P ≤ 2.18 × 10-11.''.
So it does seem likely that these represent distinct traits. We found a small number of traits that were reported more than once. But in mapping these traits to EFO terms there was a substantial amount of coarsening, where for example around 550 terms were all mapped to EFO:0007010, *drug use measurement*. 

```{r grepCKD, echo=FALSE}
xx = DBI::dbGetQuery(gwasCatSearch_dbconn(), "SELECT PUBMEDID, \"DISEASE.TRAIT\" FROM gwascatalog_metadata")
gg = grepl("chronic kidney disease", xx$DISEASE.TRAIT, ignore.case = T)
PMs = xx$PUBMEDID[gg]
##get number of EFO terms
##length(ss) is the number of terms and names(ss)[1] will be the most common trait

g1 = grepl("chronic kidney disease", gwc_df$DISEASE.TRAIT, ignore.case = T)
ss = sort(table((gwc_df$MAPPED_TRAIT_CURIE[g1])),dec=T)

gdtsub = gwc_df[g1,]

##should check that ss[1] is EFO:0004725
```



```{r studies2EFO, eval=FALSE, echo=FALSE}

manystudies = efo_df[efo_df$Direct>500,]

kable_styling(kbl(manystudies[,c(1,2,5,6)], row.names=FALSE, caption="List of EFO terms with more than 500 Studies directly mapped to them"), bootstrap_options = "striped", full_width = FALSE)

```

```{r xx2, eval=FALSE, echo=FALSE}
Sometimes the EFO terms used are very generic and hence don't provide much resolution of phenotypes.
By that we mean that they have many studies that map to them and hence are unlikely to be informative about the sort of specific questions we might want to address. These are
relatively easy to find. In Table \@ref(tab:studies2EFO) we report the set of EFO terms that have more than 500 GWAS studies mapped directly to them. The numbers in the Inherited column of Table \@ref(tab:studies2EFO) represent GWAS studies that were directly mapped to descendant terms, which are usually more specific.
```

```{r vagueTraits, echo=FALSE}
##suppress for now
tm1 = grepl("EFO:0004725", gwc_df$MAPPED_TRAIT_CURIE)

metabTraits = gwc_df[gwc_df$MAPPED_TRAIT_CURIE=="EFO:0004725", "DISEASE.TRAIT"]
zz = table(metabTraits, dnn = "Metabolic Trait")
szz = sort(zz,dec=T)[1:10]
#kable_styling(kbl(szz, row.names=FALSE, caption="10 most frequent GWAS disease traits mapping to EFO:0004725"), bootstrap_options = "striped", full_width = FALSE)
```

```{r xx, echo=FALSE, eval=FALSE, warning=FALSE}
##just a place to hide this

#One can then examine the actual traits that were mapped to each of these by querying the GWAS catalog. In Table \@ref(tab:vagueTraits) we report the 10 most frequent GWAS traits that mapped to EFO:0004725, which is *metabolite measurement*. There were `r length(metabTraits)` GWAS stuides that map to EFO:0004725 and `r length(unique(metabTraits))` unique traits.
```

### Building search corpora 

To create a searchable EFO-based corpus we think of
each EFO term as a *document*. The EFO provides a text description of
the term as well as a set of known synonyms. We show the EFO provided synonyms for `r efo_df["EFO:0003086", "Object"]`, EFO:0003086 in Table \@ref(tab:exampleEFO). 


```{r exampleEFO, echo=FALSE}
##these functions return named lists, one element for each provided EFO term
ckdSyns = getSynonyms("EFO:0003086")[[1]]
nsyn = length(ckdSyns)
##if odd add a blank
if( nsyn %% 2 ) { 
  ckdSyns = c(ckdSyns, "")
  half = (nsyn+1)/2
} else half= nsyn/2
outmat = data.frame(cbind(ckdSyns[1:half], ckdSyns[(half+1):(2*half)]))
colnames(outmat) = NULL
#ckdSyns
kable_styling(kbl(outmat, caption = paste0("Synonyms for EFO:0003086, ", efo_df["EFO:0003086","Object"])), bootstrap_options = "striped", full_width = FALSE)
```
We expand the document by including the GWAS trait label
for any GWAS study that has been mapped **directly** to that EFO term.  Our rationale
is that the curators assigned that trait description to the specific EFO term, so it 
is essentially, in their eyes, a synonym.

```{r getstudyAcc, echo=FALSE}
ckdMatches = getMatchedTraits("EFO:0003086")
#length(ckdMatches[[1]])
ckdStudies = gwc_df[gwc_df$MAPPED_TRAIT_CURIE=="EFO:0003086",]
#ckdStudies$STUDY.ACCESSION

```

For clarity, we create one *document* for each EFO term. The document has three components, the term itself we call the *object*, second a set of zero or more *synonyms* provided by EFO and third a set of zero or more *matches* which are GWAS study traits that have been mapped directly to the EFO term.

Searching for relevant GWAS studies using this corpus one first obtains all EFO terms where the tokenized search string matches at least one of the *object*, *synonyms* or *matches* for that term.  The desired GWAS studies are those that have either matched directly to that EFO term, or are inherited by that term. 

Suppose our search string had been *chronic kidney disease*, then we would have found 
the term EFO:0003086, which has as its text description *chronic kidney disease*. 
There were `r length(ckdMatches[[1]])` GWAS catalog studies that were mapped **directly** to this EFO term and `r efo_df["EFO:0003086", "Inherited"]` terms that were inherited, i.e. mapped to more specific EFO terms than EFO:0003086, which could be subtypes.

```{r efotcex, echo=FALSE, message=FALSE, warning=FALSE}
##build up one document example, for EFO:0003086
efoidx=grep("EFO:0003086", efo_df$Subject)
ans = list(Object = efo_df$Object[efoidx], Synonyms=getSynonyms("EFO:0003086"), 
           Matches =getMatchedTraits("EFO:0003086"))
```

##FIXME: Vince here is a place where a picture of the EFO graph could be useful

#### A search corpus based on GWAS Traits.

```{r GWCexp1, echo=FALSE}
GWCexp1 = gwc_df[which(gwc_df$STUDY.ACCESSION=="GCST90083943"),]
EFO1 = GWCexp1$MAPPED_TRAIT_CURIE
EFO1Syns = getSynonyms(EFO1)
GWCexp1$EFOSynonyms = paste(unlist(EFO1Syns), collapse="; ")

kable_styling(kbl(t(GWCexp1), align='lc', caption="Study GCST90083943", col.names=""), bootstrap_options = "striped", full_width = FALSE)

```

To create this searchable corpus we made documents, one for each GWAS study, by assembling the GWAS trait (DISEASE.TRAIT), the EFO mapped trait (Trait) and all EFO provided synonyms (Synonyms). In this case the ontology is being used to provide a standardized trait name and its synonyms. The relationships within the ontology are not used. The corpus can be searched and studies that match the search term retrieved. 
In Table \@ref(tab:GWCexp1) we show the metadata associated with the study GCST90083943 as well as the synonyms for `r EFO1`.

```{r exgwctc, echo=FALSE, warning=FALSE, message=FALSE}
idsx1 = grep("GCST90083943", gwc_df$STUDY.ACCESSION)
ans = list(DISEASE.TRAIT = gwc_df$DISEASE.TRAIT[idsx1], Trait=gwc_df$MAPPED_TRAIT[idsx1], Synonyms=getSynonyms(gwc_df$MAPPED_TRAIT_CURIE[idsx1]))

```

The same set of textual descriptions are used for constructing both this corpus and the EFO corpus described above.  These are the original set of traits as defined by the authors of the studies and the traits and synonyms for all EFO terms that were mapped to.  So a search based on either one of these should yield the same set of EFO terms and GWAS catalog studies. We find that it is useful to have both.

```{r GWCexp2, echo=FALSE}
##find all the studies that map to leukemia
htl = search_features(gwc_tc, query="leuk*")
htlf = addField2Hits(htl, gwc_tc)
tt = table(htlf$hits$field)

kable_styling(kbl(tt, caption="Which part of the GWAS *document* matched the search term leuk*", col.names = c("Term", "Count")) , bootstrap_options = "striped", full_width = FALSE)
```
```{r testcode, echo=FALSE, warning=FALSE, message=FALSE}

queries = data.frame(label = c('Onset','Duration'),
                     query = c("age AND (onset OR begin)", "duration NOT sleep*"))
t1h = search_features(gwc_tc, query=queries$query, code=queries$label)
t1hf = addField2Hits(t1h, gwc_tc)
count_tcorpus(gwc_tc, hits=t1hf)
```

As shown in Table \@ref(tab:GWCexp2) it is also possible to identify which of the three parts of the *document* described above have matches.
There were `r length(unique(htlf$hits$doc_id))` unique documents that
were identified. We can see that there were `r tt["DISEASE.TRAIT"]`
cases where a match was to the DISEASE.TRAIT, `r tt["Synonyms"]` matches
to synonyms provided by the ontologies, and `r tt["Trait"]` matches to
the text description used by the ontology for the trait. 

## Analysis: Searching for diseases and traits

There are a number of tasks that one might be interested in using the GWAS catalog to aid in.
Finding all diseases of a specific type, say disorders of the kidney, or infectious disease and finding diseases that relate to some organ, system or part of the body.  Having traits mapped to an ontology, such as EFO, should help with these tasks.  We will examine how these tasks can be supported using the EFO ontology together with the search corpora we describe above. This process basically integrates a dictionary search with the structured information in an ontology.

```{r diseaseN, echo=FALSE, message=FALSE, warning=FALSE}
didx = grep("EFO:0000408", efo_df$Subject)

```

We first consider the question of whether the GWAS traits are mapped to ontology terms that represent diseases.  The EFO CURIE for *Disease* is EFO:0000408 and all diseases should be children of that node.
We find that `r efo_df$Direct[didx]` GWAS studies were mapped directly to this node while `r efo_df$Inherited[didx]` were inherited. Since there are `r length(unique(gwc_df$STUDY.ACCESSION))` different GWAS studies only a small fraction of the studies are mapped to diseases.

We last point out that investigator assigned disease traits in the GWAS Catalog metadata may also cause some challenges.  We searched each disease traits for any occurance of the string *chronic kidney disease*, ignoring capitalization, and found `r sum(gg)` matches which correspond to `r length(unique(PMs))` different PUBMEDIDs. And these studies are mapped to a wide variety of EFO terms, `r length(ss)` different terms, with `r ss[1]` mapping to `r names(ss)[1]`, which is *metabolite measurement*.


We will look for GWAS
studies associated with *infectious disease*.  One reason for this choice is that it is quite a challenging problem if one attempts it without using an ontology.  The main reason for that is that there are many infectious diseases and most of them do not have either *infectious* or *disease* in their name.  So without an ontology to guide us, we would need to have some dictionary of infectious diseases to start with. [FIXME: should we try to find one? MeSH or something similar?]

### Strategy One: Use the EFO ontology structure


```{r tryChildren, echo=FALSE}
#first find the EFO term for infectious diseases
gg = grep("^infectious disease$", efo_df$Object, ignore.case=TRUE )
##efo_df[gg,]

##now look at all the more specific terms
z2 = EFOdescendants("EFO:0005741")
##length(z2)

efo_id = efo_df[z2,]
##d
## how many studies map directly to one of these terms
knitr::kable(t(table(efo_id$Direct)), caption="Number of studies mapping to descendants of EFO:0005741") |> kable_styling()
## one has 249 - and it is COVID, which makes some sense
# efo_id[which.max(efo_id$Direct),]

##drop those with 0 mapping to them
efo_ids = efo_id[efo_id$Direct>0,]

##below shows how to find the GWAS studies that map
dirMap = resources_annotated_with_term(efo_ids$Subject, include_subclasses=FALSE)
#dim(dirMap)
```

Here we describe how to use these tools start with a general type of disease, say infectious
disease, and then try to find all GWAS studies that report on that class of 
diseases. First we identify the node in the EFO ontology that
corresponds to *infectious disease*, which has CURIE EFO:0005741. All infectious diseases should be
annotated in the EFO as descendants of that EFO term. Currently there are `r length(z2)` descendants and a total of `r efo_df[gg, "Direct"] + efo_df[gg, "Inherited"]` GWAS studies annotated at EFO:0005741 or one of its descendants. We find
that `r efo_df[gg,"Direct"]` were directly mapped to EFO:0005741 while `r efo_df[gg,"Inherited"]` where inherited. The term in Table \@ref(tab:tryChildren) with over 200 studies is `r efo_id[which.max(efo_id$Direct),"Object"]`,

 FIXME: need some way to look at the trait descriptions for the EFO terms that are descendants of
 EFO:000571 and see if they are sensible - perhaps compare leaf vs internal nodes?
 
### Strategy Two: Search based on the GWAS catalog corpus

In this case we will search the corpus based on the GWAS catalog. One might want to experiment with different queries to try to obtain a good trade off between sensitivity and specificity.  We found that using *infectious diease* as our query missed a number of GWAS studies that had *infection* in them and opted for the term *infect\**. If a list of all infectious diseases was available a different strategy would be to search for each disease in that list and then merge the results.   Recall that when we constructed the GWAS catalog corpus we used the text descriptions of the GWAS trait as well as the trait name and the synonyms that correspond to the EFO term(s) this study was directly mapped to.

```{r IDgwctc, echo=FALSE}
qID = "infect*"
htsID = search_features(gwc_tc, query = qID)
htsID = addField2Hits(htsID, gwc_tc )
hitsIDasDT = hits2DTGWC(htsID, gwc_df, gwc_tc)
```

We found `r{NROW(hitsIDasDT` and in Table \@ref(tab:checkIDmatches) we report where in the *document* matches have occurred, if a match was in more than one part of the *document* then both are reported,  we see that the majority of matches are to synonyms.

```{r checkIDmatches, echo=FALSE}
idTT = table(hitsIDasDT$field)
kable_styling(kbl(idTT, caption="Which part of the text matched our search term.", col.names = c("Term", "Count")) , bootstrap_options = "striped", full_width = FALSE)
```



```{r compID, echo=FALSE}
vv = setdiff(hitsIDasDT$Study, dirMap$STUDY.ACCESSION )
v2 = setdiff(dirMap$STUDY.ACCESSION, hitsIDasDT$Study)

#DT::datatable(hitsIDasDT[match(vv, hitsIDasDT$Study),])

vvGWC = gwc_df[match(vv, gwc_df$STUDY.ACCESSION),]
v2GWC = gwc_df[match(v2, gwc_df$STUDY.ACCESSION),]

##see if the ancestor of the first term is under infectious disease - and it is not
check1 = EFOancestors(vvGWC[1, "MAPPED_TRAIT_CURIE"])
#match("EFO:0005741", check1)

##look at the other list - and we see that these are infectious diseases
```

We can ask if this second approach found any GWAS studies that the first one did not.
And further look at those, to see if any are likely to be about infectious diseases. We
will also check to see if the first method found studies that the second approach did not.
In this case this second approach found `r length(vv)` studies that were not found by our first approach, while the first approach found `r length(v2)` studies that were not found by the second.

One of the studies found by this method that were not found using the EFO structure was `r vv[1]` which has the disease trait `r vvGWC[1, "DISEASE.TRAIT"]`. And the associated EFO term is `r vvGWC[1, "MAPPED_TRAIT_CURIE"]` and while it is not an infectious disease one of its synonyms is "pseudomonas aeruginosa, susceptibility to chronic infection by, in cystic fibrosis", which explains why it was found using our second method.  Unfortunately many of the synonyms reported in EFO are not actually synonyms and this will cause matches where there should not be any.

Examining the other list, those found by our first method, but not by our second we see that these are mainly infectious diseases where the word infect is not included in the name of the disease or any of its synonyms.

### Strategy three: Use the EFO corpus directly

Above we first started by searching for *infectious disease* in the EFO ontology and then found
all GWAS studies that mapped directly to either that node, or one of its descendants. We contrasted that with the strategy of searching the corpus based on the GWAS catalog studies for the terms infectious and disease.  A third approach is to search the EFO corpus for *infectious disease*.  However, this approach will simply search the same text strings, although arranged differently. 

```{r otherterms, echo=FALSE, eval=FALSE}
##Not sure we need to, or want to evaluate this any more
qIDefo = "Infectious AND disease"
htsID2 = search_features(efo_tc, query = qIDefo)
hitsID2asDT2 = hits2DT(htsID2, efo_df, efo_tc)
##DT::datatable(hitsasDT2[1:100,], escape=FALSE, rownames=FALSE)

```

Now we can compare the EFO categories and see if we found anything new and also examine
the GWAS studies. A quick visual inspection of the EFO categories found by this method
and not the first, suggests that the first approach did seem to get everything useful.

```{r method3comp, echo=FALSE, eval=FALSE}
efoIDS3 = row.names(hitsID2asDT2)
setd3 = setdiff(efoIDS3, row.names(efo_id))
#efo_df[setd3,"Object"]

```

We end this section by creating a Manhattan plot based on the studies that were identified using the first strategy. 

```{r getresources, echo=FALSE}

IDassoc <- granges_from_study(dirMap$STUDY.ACCESSION)
IDassoc2 = IDassoc[IDassoc$P.VALUE<1e-8,]

#NROW(IDassoc2)
##how many are reported more than once...
#sum(duplicated(IDassoc2$SNP_ID_CURRENT))
##and produce a plot
```

```{r IDmanh, echo=FALSE, plot=TRUE, fig.align='center', fig.width=7, fig.cap="Manhattan Plot for Infectious Disease", warning=FALSE, fig.height=5, fig.path="figure/"}
pp = variants_from_study(dirMap$STUDY.ACCESSION)
pp = pp[IDassoc$P.VALUE<1e-8,]

#pp = ggmanh::manhattan_data_preprocess(IDassoc2, pval.colname="P.VALUE")
simple_ggmanh(pp, y.label="-log10 p",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

##ggmanh::manhattan_plot(pp, plot.title="Infectious Diseases", y.label="-log10 p")
```

## Search for GWAS studies by Disease Location

In recent versions of the EFO the curators have begun to include disease location information.
This information is encoded using the UBERON ontology. To use this functionality you would first identify the UBERON CURIE that represents the location of interest.  In the code below we look for diseases associated with the lung.  

FIXME: Vince - any sort of EFO graph that we could use to represent the disease categories associated with lung?

The CURIE for lung in the UBERON ontology is UBERON:0002048.  We identify
all EFO terms that refer to this CURIE as their disease location. And then use corpusTools to create a document term matrix (DTM) to represent the frequency of different words in the subcorpus based only on terms associated with UBERON:0002048.  We can then visualize this in Figure \@ref(fig:subsetLung).

RAFAELFIXME - can there be sub-terms? and do we need some tools to roll up?

```{r uberon, warning=FALSE, echo=FALSE}
lungU = "UBERON:0002048"
lungEFO = grepl(lungU, efo_df$DiseaseLocation)

##get the actual EFOs -
EFOsforLung = efo_df[lungEFO, "Subject"]
##now get the set of GWAS study IDs - this is DF/subset of gwc_df
GWLungstudies = resources_annotated_with_term(EFOsforLung)

efo_tc$meta$Lung = lungEFO

efo_lung = efo_df[lungEFO,]
##FIXME - here it would be good to get some form of graph completion of the EFO labels

```

In the code below we subset first to the set of terms that have an association with the lung and then to the set of tokens that have more than 5 occurrences. And finally we create a word cloud based on the frequency of the tokens that remain.  These are based on occurance in the EFO corpus.

```{r subsetLung, fig.cap="Word Cloud based word frequency in the EFO labels of those EFO terms that have disease location  UBERON:0002048. The data were subset to only include tokens that have at least 5 occurrences.", warning=FALSE, echo=FALSE}

efo_tcLung = subset(efo_tc, subset_meta = Lung == TRUE)
efo_tcLung2 = subset(efo_tcLung, subset=freq_filter(token, min=5))
dtmLung = get_dtm(efo_tcLung, "feature")
csums = apply(dtmLung, 2, sum, na.rm=T)
#sort(csums/sum(csums),dec=T)[1:10]
dtm_wordcloud(dtmLung)
```



### Exploring the GWAS metadata

Once we have a set of GWAS studies that we are interested in we can access the metadata for the implicated variants such as likely gene, the genome coordinates, the p-value and effect size to name a few of the attributes that are provided.  These data are large and we use a  database as the main tool for storing and accessing the data. The data can be read into standard Bioconductor objects, such as those for storing and manipulating genetic coordinates and users can easily produce Manhattan plots, or identfy genes that are implicated.  In Table  \@ref(tab:lookatHits) we report the ten most frequent mapped genes for the lung studies.



```{r from_studyfuncs, echo=FALSE}
##there should only be one mapped trait curie - not sure what
##happens here with the studies that have multiple EFO terms
tab = granges_from_study(acc=GWLungstudies[,"STUDY.ACCESSION"])


```


```{r lookatHits, eval=TRUE, echo=FALSE}
kable_styling(kbl(t(sort(table(tab$MAPPED_GENE),dec=TRUE)[1:10]), caption="Most frequent mapped genes in GWAS lung studies." ), bootstrap_options = "striped", full_width = FALSE)

```

We can then create a Manhattan plot of the variants associated with the selected studies.  There are many different ways in which one can explore these data. The metadata is quite extensive and one might also want to subset, or color variants by disease type.

```{r lungmanh, plot=TRUE, echo=FALSE, fig.align='center', fig.width=7, fig.cap="Manhattan Plot for Disease Location Lung", warning=FALSE, fig.height=5, fig.path="figure/"}
vars = variants_from_study(GWLungstudies$STUDY.ACCESSION)
##drop those with a p-value > 1e-8,
vars = vars[vars$`P-VALUE` < 1e-8, ]
simple_ggmanh(vars, y.label="-log10 p", plot.title = "Diseases of the Lung",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

```


## Challenges

While the GWAS catalog has provided mappings to EFO there remain some substantial challenges
for those attempting to use it to find and organize their searches.  These include vague EFO assignments where hundreds of phenotypes have been mapped to an EFO term, papers that give rise to thousands of GWAS studies, inaccuracies and inconsistencies in the EFO, and a lack of clarity on the explicit rationale for mapping a trait to an EFO term.  We give examples of each of these.


### Mapping and Interpreting

```{r efoCKD, echo=FALSE}
q1 = "chronic AND kidney AND disease"
hts = search_features(efo_tc, query = q1)
hts = addField2Hits(hts, efo_tc )
hitsasDT = hits2DT(hts, efo_df, efo_tc)
```

```{r gwcCKD, echo=FALSE}
htsgwc = search_features(gwc_tc, query=q1)
htsgwc = addField2Hits(htsgwc, gwc_tc )
hitsgwcasDT = hits2DTGWC(htsgwc, gwc_df, gwc_tc)
```

We noted above that the term EFO:0003086 has text description of "chronic kidney diseases".
When we 

In this example we are examine kidney function traits, and in particular we
will search for *chronic AND kidney AND disease* in the EFO-based corpus. This string should find EFO terms that have matchs to all three words that are close to each other.

```{r gwasCatPubMed, echo=FALSE}
q1 = "chronic AND kidney AND disease"
hts = search_features(efo_tc, query = q1)
hts = addField2Hits(hts, efo_tc )
hitsasDT = hits2DT(hts, efo_df, efo_tc)
```

```{r DT, fig.cap="CKD EFO Hits", echo=FALSE}
knitr::kable(hitsasDT[1:10,], escape=FALSE, row.names=FALSE, caption="CKD EFO Hits (first 10)")
##DT::datatable(hitsasDT[1:100,], escape=FALSE, rownames=FALSE, caption="CKD EFO Hits (first 100)")
```

Surprisingly we find that there were `r nrow(hitsasDT)`
different EFO terms that have a hit. This seems very large to
understand what is happening we start by looking 
code chunk we explore some of the issues. If we consider the EFO node for
*kidney disease* we anticipate that all different forms of kidney
disease will be descendants of that node. And we expect that all `r nrow(hitsasDT)`
hits will map to some one of those nodes.

```{r ckdEFO, echo=FALSE}
kdEFO = grep("^kidney disease$", efo_df$Object, ignore.case = TRUE)
#efo_df[kdEFO,]
ackd = EFOdescendants("EFO:0003086")
```

There are only `r length(ackd)` EFO terms that are more specific than the term *kidney disease* and the total number of GWAS studies that mapped directly to EFO:0003086 or were inherited we only get `r efo_df["EFO:0003086","Direct"] + efo_df["EFO:0003086","Inherited"]` GWAS studies.  This suggests that there are a very large number of studies where the GWAS trait contains the phrase *chronic kidney disease* that were not mapped to any EFO term directly related to kidney disease. And unfortunately that suggests that to find those GWAS catalog studies we will need to rely on string matching which is not the best approach.


```{r msmt, echo=FALSE}
msmtdesc = EFOdescendants("EFO:0001444")
dtandmsmt = intersect(msmtdesc, row.names(hitsasDT))
#length(dtandmsmt)
```

A little more exploring suggests that many of the terms are actually mapped to the descendants of the *measurement*, EFO:0001444, node. Specifically we find that there are `r length(dtandmsmt)` EFO terms that were identified by our search which are annotated as measurements, a number far larger than those that align to disease.  Which makes searching for terms related to any specific disease somewhat more challenging, as the organizational principles for this part of the EFO are not suitable for finding related diseases and it appears the analyst is left with string matching as the only viable option.

This means that the bulk of the EFO terms are showing up due to finding
the text string in the name of some other GWAS catalog trait and 
that trait was not mapped to any EFO label related to kidney disease. 


We next query the GWAS catalog traits.
The dataframe `gwc_df` contains the names of all traits in the `DISEASE.TRAIT` column.

```{r gwcmetadtable, echo=FALSE}

gg = grep("chronic kidney disease", gwc_df$DISEASE.TRAIT, ignore.case=T)


gl= grepl("chronic kidney disease", gwc_df$DISEASE.TRAIT, ignore.case=T)
gwsub= gwc_df[gl,]
r1 = grep("measurement",gwsub$MAPPED_TRAIT,  ignore.case = TRUE)
#length(r1)

##gwc_df[gg[1:10], "DISEASE.TRAIT"]
ckdPMIDs = GWCATID2PMID(unique(gwc_df$STUDY.ACCESSION[gg]))
pmtab = sort(table(ckdPMIDs), dec=TRUE)
#pmtab[1:10]
```

And now we see that there are in fact `r length(unique(gwc_df[gg, "STUDY.ACCESSION"]))` terms that have been labeled this way. Further, if we look at the PUBMED IDs associated
with these traits we see that there are `r sum(pmtab>1000)` papers that
report more than 1000 traits. Given the PubMed IDs it is easy to find
the papers. The paper associated with PMID 35870639 which reports more than 
6,700 studies is titled: Identification of 969 protein quantitative trait loci in an African American population with kidney disease attributed to hypertension @surapaneniIdentification969Protein2022.


We end this example by creating a Manhattan plot based on the studies that contain the string *chronic kidney disease*.

```{r ckdmanh, plot=TRUE, echo=FALSE, fig.align='center', fig.width=7, fig.cap="Manhattan Plot for CKD", warning=FALSE, fig.height=5, fig.path="figure/"}
vars = variants_from_study(gwc_df$STUDY.ACCESSION[gg])
##drop those with a p-value > 1e-8,
vars = vars[vars$`P-VALUE` < 1e-8, ]
simple_ggmanh(vars, y.label="-log10 p", plot.title = "Chronic Kidney Disease",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

```

### Bad synonyms

Another potential issue is that the EFO provided synonyms might be inaccurate. Further investigation of those issues is also recommended. FIXME: maybe Rafael has some ideas.

For `r efo_df["EFO:0000685", "Object"]` the reported synonyms include *proliferative arthritis* and
*Rheumatic gout*. However, proliferative arthritis refers to a condition characterized by the excessive growth of synovial tissue in the joints, leading to inflammation and joint damage. Rheumatoid arthritis, on the other hand, is an autoimmune disease where the immune system mistakenly attacks the joints, causing inflammation, pain, and joint damage. And *Rheumatic gout* is a somewhat archaic term for gout. Including these as synonyms will result in linking traits that should not be linked.



```{r semnet, warning=FALSE, message=FALSE, echo=FALSE}
gwc_tc$preprocess(min_docfreq = 50, remove_stopwords = T, remove_numbers = T)

g = semnet_window(gwc_tc, 'feature')
gb = backbone_filter(g, alpha = 0.001, max_vertices = 100)
plot_semnet(gb)

fa = feature_associations(gwc_tc, 'feature', query = 'chron*')
fa = feature_associations(gwc_tc, 'feature', query = 'chronic AND kidney')

fa = feature_associations(gwc_tc, 'feature', query = "leuk*")
```

#### Other Hierarchies

While much of the EFO of interest is organized around *diseases* (EFO:0000408) there is also
a *measurement* (EFO:0001444) node, with about 8,700 terms nested under it. These tend to be measurements of some of the very same diseases that are arranged under the *disease*  node. Unfortunately, there is no easy way to go between the two. They are not related in the EFO hierarchy and we have been unable to find an organizing principle.  Users likely will want to search under both of these high level nodes for matches to any disease, gene or phenotype of interest.   For example, in our infectious disease example we see that there are nodes 
*susceptibility to infectious disease measurement* (EFO:0008422) and
*infectious disease biomarker* (EFO:0006843) which may also be relevant.
FIXME: report the number of studies that. align under the measurement node

##Results and Discussion

The results of our exploration are encouraging but also exhibit the challenges that remain. The mapping of phenotypes to an ontology is extremely helpful in providing a more specific vocabulary that can be used to harmonize across studies.  The terms or nodes in the ontology provide a means of associating synonyms and they can then easily be incorporated into searches.  Ontologies provide some means of versioning and hence of updating and improving the ontology.  And as we have noted that additional information, such as disease location, can also be documented and then subsequently used in any analyses.

While ontologies provide mechanisms for describing complex relationships their practical application to support specific scientific investigations remains challenging.  There are no clear mechanisms to describe the processes used.  For example we have shown that in many cases the curators chose to map phenotypes to entries in the *measurement* portion of the EFO ontology, which may be appropriate for many situations, but as our focus is in identifying diseases this decision appears to leave us relying on syntax and string matching.
  The ontologies are incomplete, which is to be anticipated, but that incompleteness leads to compromises in the labeling of phenotypes for which there are no standards for describing.
  We anticipate that further refinements such as making use of other natural language processing tools such as name-entity-recognition (NER) may be useful in further identifying phenotypes that reference specific diseases, genes, drugs.
  Our ability to integrate and synthesize data across studies remains hampered by other issues as well.  In most situations we would like to be able to address whether or not studies were based on reasonably similar cohorts of individuals, and in some cases whether or not the inclusion and exclusion criteria were similar.  This of course remains a challenge across much of medical and epidemiological data reporting.
  
  
## Citations



### Appendix

The GWAS catalog data are obtained from xhttps://www.ebi.ac.uk/gwas/
and transformed according to the descriptions in https://github.com/ccb-hms/GWASCatalogSearchDB,
which gives explicit details on which files were obtained, and the dates on which they
were obtained.
