---
title: "Paper"
author: "RG"
date: "2023-06-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("gwasCatSearch")
library("DT")
```

## Example 1
Example in kidney function traits (Li et al., 2020) PMID 32764137

The query

(glomerul* AND filtration) OR urate OR (chronic AND kidney AND disease) OR (urea AND nitrogen) produces, via search_features and hits2DT, 7 ontologic tags:


```{r }

q1 = "(glomerul* AND filtration) OR urate OR (chronic AND kidney AND disease) OR (urea AND nitrogen)"
hts = search_features(efo_tc, query = q1)
hitsasDT = hits2DT(hts, efo_df, efo_tc)
DT::datatable(hitsasDT, escape=FALSE, rownames=FALSE)
```

## Mapping to the Ontologies and finding the roll-ups


```{r pressure, echo=FALSE}
ematch = resources_annotated_with_term("EFO:0009909", include_subclasses=FALSE)
dim(ematch)
allmatch = resources_annotated_with_term("EFO:0009909", include_subclasses = TRUE)
dim(allmatch)
DT::datatable(head(allmatch,n=20))
```

Now we can use the `variants_from_study` function to obtain information about the GWAS hits such as the `rsID`, risk allele and effect size.

```{r lookupSNP}
lu1 = variants_from_study(allmatch$STUDY.ACCESSION[1],gwcat=gwascat_2023_06_24)
lu1$MAPPED_GENE

```

And then we can visualize the variant context using `view_variant_context`.

```{r plotContext, echo=FALSE}
view_variant_context(chr=lu1$CHR_ID[1], pos=lu1$CHR_POS[1], radius=5e+05,
                     gwcat=gwascat_2023_06_24, main="Manhattan Plot")
```
