---
title: "Paper"
author: "RG"
date: "2023-06-23"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Paper}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("gwasCatSearch")
library("DT")
```

## Introduction

Larger and more complex biomedical data sets are becoming increasingly
available and they provide substantial opportunities for discovery.
Realizing the potential for discovery is challenged by the complexity of
the data, the inherent relationships between phenotypes and the need for
data analysts to be able to navigate some of the complexities of both
biological and medical terminologies. In this paper we explore using
corpus search tools (as exemplified in the `corpustools` package
@citeCorpustools) to help search through the set of traits that are
reported by the GWAS catalog. These traits have been hand curated by the
EBI and mapped to terms in the Experimental Factor Ontology (EFO,
@citeEFO). However, the tools available for searching these data are
mainly variants of regular expression matching, which as very limited
expressivity for the complex nature of medical and biological terms.

The GWAS Catalog (cite gwascatalog) reports on a set of curated GWAS
association studies. Since 2010, delivery and development of the Catalog
has been a collaborative project between the EMBL-EBI and NHGRI. As of
2024-01-30, the GWAS Catalog contains 6715 publications, 571148 top
associations and 67322 full summary statistics. GWAS Catalog data is
currently mapped to Genome Assembly GRCh38.p14 and dbSNP Build 156.

Among the curation activities is a process of mapping the traits that
are studied to the EFO (cite EFO). In part this allows one to search for
studies about a particular disease, but for a variety of reasons this
search can be quite complicated. Here we report on some tools and
approaches that can be used to help find GWAS studies according to
specific criteria. Our software is contained in the `gwasCatSearch` package (@citegwasCatSearch) 
which is open source and freely available.

## Methods

### Data
We obtain data from the GWAS Catalog and transform it into a set of tables using tools in the GWASCatalogSearchDB (https://github.com/ccb-hms/GWASCatalogSearchDB). This processing produces
a number of tables that can be accessed via a SQL database; we use SQLite (@cite(SQLite)). These tables include the raw data from the EBI, several tables organized around the Experimental Factor Ontology (EFO), the Uber-anatomy ontology (UBERON, https://www.ebi.ac.uk/ols4/ontologies/uberon).
An ontology is a set of concepts and categories that describe a subject area or domain of knowledge that expresses their properties and the relations between them. Ontologies form one of the bases for knowledge engineering and representation and are important in understanding and referencing biomedical entities. One of the challenges that needs to be addressed is the fact that the same concept might usefully appear in different ontologies and users will need tools to help identify similar concepts. A second challenge is that to be useful the ontologies must be accessible and typically that means using the web or a database implementation. There are two useful constructs that help with this, the compact uniform resource identifiers (CURIEs), and the internationalized resource identifiers (IRIs)  which in ontology space are used as *names* terms in the ontology. A CURIE has three parts: a prefix; a delimiter; and a unique local (to the ontology described by the prefix) identifier from the given nomenclature. The CURIE EFO:0005305 which represents the concept *atrioventricular node disease* has a prefix of EFO, a delimiter of a colon (:) and a unique local identifier of 0005305.  The corresponding IRI for that term is http://www.ebi.ac.uk/efo/EFO_0005305, which gives a web location that can be examined by a user.

As part of the processing we construct two dataframes, one based on the EFO terms `efo_df` and the other based on the GWAS catalog studies `gwc_df`. These are described in Table 1. We create two different corpora for searching. One based on the EFO and
the relationships inherent in it, `efo_tc`, and the second based on the GWAS
catalog studies, `gwc_tc`.

The EFO dataframe has one row for each term in the EFO ontology and columns labeled `Subject`: the corresponding EFO CURIE; `Object`: the textual description of the trait; `IRI`: the IRI for that term; `DiseaseLocation`: the UBERON CURIE for the location of the `trait` if there is one; `Direct`: the number of GWAS catalog studies that were mapped directly to that trait; `Inherited`: the number of traits that were mapped to terms that are more specific (according to the ontology) than the `Object`.

Since a GWAS study can be directly mapped to more than one EFO term we have made our
dataframe have one row for each unique combination of GWAS catalog study ID and EFO term.
The GWAS catalog dataframe has one row for each GWAS catalog study and columns labeled `STUDY.ACCESSION`: the GWAS catalog study identifier; `DISEASE.TRAIT`: the GWAS catalog text description of the trait; `MAPPED_TRAIT`: the textual description of the trait from EFO that this study was *directly* mapped to; `MAPPED_TRAIT_URI`: the URI for the mapped trait; `MAPPED_TRAIT_CURIE`: the CURIE for the mapped trait.

Each GWAS study has a specific trait that was studied and that is
reported as a text string. That trait is mapped to a specific term in
the EFO ontology, and we refer to those as **direct** mappings. The terms in the EFO ontology
are related to each other via a directed acyclic graph. By construction there is a single root node. More specific terms, such as the relationship between cancer and pancreatic cancer are encoded with pancreatic cancer being the **child** of cancer,
while cancer is the **parent** of pancreatic cancer.  The descriptions can become more refined, and we define the **descendents** of a node to be the set of child nodes, where one recurses down to the point where a node has no children.  Such a node is referred to as a *leaf*.  Similarly **ancesters** are the parents, again applied recursively to the root node of the ontology.  Our software contains tools to obtain the children, descendents, parents, and ancestors as well as to perform other operations on the ontology. 

Tools for representing and organizing biomedical entities and concepts are key components of
any system that process disease, medical or epidemiological data.  There are many challenges among them is the need to have some way to have two different systems align on some concepts. On the seminatic web Uniform Resource Identifiers (URIs) are used to reference named entities. So URI basically give a web address where more details on the concept can be found.

Given the structure of the ontology we can introduce the concept of an *inherited* mapping. A mapping at a node E is inherited if it was a direct mapping to some descendant of E.  Basically this is
providing a way to find, and hence examine, the EFO terms that map to more specific versions of an EFO term.  So, while we might want to know the GWAS catalog traits that map to cancer, we may also want to find the traits that map to more specific types of cancer.  This concept allows us to do that in a rigorous way.

### The EFO Corpus

To create a searchable EFO-based corpus we think of
each EFO term as a *document*. The EFO provides a text description of
the term as well as a set of known synonyms. We expand that by including the GWAS trait 
for any GWAS study that has been mapped **directly** to that EFO term.  Our rationale
is that the curators assigned that trait description to the specific EFO term, so it 
is essentially, in their eyes, a synonym.

For clarity, there is one document for each EFO term. The term itself we call the *object*,
that is augmented by a set of zero or more *synonyms* provided by EFO and then additionally
augmented by a set of zero or more *matches* which are GWAS study traits.  Below
we examine one node, EFO:0003086, which is labeled as chronic kidney disease.
FIXME: we probably want some way to get the GWAS catalog 
```{r exampleEFO}
##these functions return named lists, one element for each provided EFO term
ckdSyns = getSynonyms("EFO:0003086")
ckdSyns
ckdMatches = getMatchedTraits("EFO:0003086")
length(ckdMatches[[1]])
ckdStudies = gwc_df[gwc_df$MAPPED_TRAIT_CURIE=="EFO:0003086",]
ckdStudies$STUDY.ACCESSION

```

At this point all we have done is asked which EFO terms match our
search. And by match our search we mean that one or more of the elements
of the text search is found in either the name of the EFO term, the EFO
provided synonyms for that term, or the names of the set of GWAS catalog
studies that the EBI has annotated with that same term.

## Example 1

Here we use the EFO (ontology) labels as the basis for constructing our
searchable corpus.

Suppose we are interested in kidney function traits, and in trying to
find GWAS studies that report on these traits.

```{r gwasCatPubMed}
q1 = "chronic AND kidney AND disease"
hts = search_features(efo_tc, query = q1)
hts = addField2Hits(hts, efo_tc )
hitsasDT = hits2DT(hts, efo_df, efo_tc)
nrow(hitsasDT)
DT::datatable(hitsasDT[1:100,], escape=FALSE, rownames=FALSE)
```

We could start with a search of EFO for chronic kidney disease. After
running the code above we find that there were `r nrow(hitsasDT)`
different EFO terms that have a hit, at least according to the methods
we have used. This seems somewhat too large to be helpful. In the next
code chunk we explore some of the issues. First we can look to see where
the matches to *chronic kidney disease* are occurring. We can look at
the `hits` object, and also at the `datatable`, the former gives one
record for each *hit*, while the latter gives an indication of what type
of string was matched on. The `Object` is the text description of the
EFO term, the `Synonyms` are the EFO provided synonyms for the term, and
the `Matches` are the GWAS catalog trait descriptions for traits that
mapped directly to that EFO term. For example, for `r hitsasDT$EFO[1]`
the string *chronic kidney disease* was matched only to other GWAS
catalog traits (since only matches were found).

```{r why7K}
table(hts$hits$field)
table(hitsasDT$field)
```

This means that the bulk of the EFO terms are showing up due to finding
the text string in some other GWAS catalog trait. So there must be a lot
of GWAS catalog traits that contain that string. So we can try to find
out about this by querying the `gwc_df` data.frame.

```{r gwcmetadtable}
gg = grep("chronic kidney disease", gwc_df$DISEASE.TRAIT, ignore.case=T)
length(gg)
gwc_df[gg[1:10], "DISEASE.TRAIT"]
ckdPMIDs = GWCATID2PMID(gwc_df$STUDY.ACCESSION[gg])
pmtab = sort(table(ckdPMIDs), dec=TRUE)
pmtab[1:10]
```

And now we see that there are in fact `r length(gg)` terms that have
been labeled this way. Further, if we look at the PUBMEDIDs associated
with these traits we see that there are `r sum(pmtab>1000)` papers that
report more than 1000 traits. Given the PubMed IDs it is easy to find
the papers.

We can then create

```{r ckdmanh, plot=TRUE}
vars = variants_from_study(gwc_df$STUDY.ACCESSION[gg])
simple_ggmanh(vars, y.label="-log10 p", plot.title = "Chronic Kidney Disease",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

```

So from the datatable we can see that the EFO label for chronic kidney
disease is EFO:0003884 and that there were 71 studies that mapped
directly to that term, and 47 that are associated with that term, but
likely were mapped to more specific types of chronic kidney disease
(FIXME: is it easy to figure out what those reall were mapped to?). Next
we want to look at the studies that were

```{r findresources}
#find the resources annotated with this term
dirMap = resources_annotated_with_term("EFO:0003884", include_subclasses=FALSE)
#see if any papers are giving rise to multple studies
sort(table(dirMap$PUBMEDID), dec=TRUE)[1:10]
## we see one that has 10 studies PMID 34662886
##when we retrieve it, we see that there are a lot of GWAS studies reported
vv = PubMed2GWASCAT("34662886")
dim(vv)
```

## Vague Study Definitions

Now, one challenge might be that some of the EFO terms are just too
vague, and have too many studies that map to them. We report on those
with more than 500 directly mapping studies in the code below.

```{r studies2EFO}

manystudies = efo_df[efo_df$Direct>500,]
DT::datatable(manystudies[,c(1,2,5,6)], rownames=FALSE)


```

## Approach 2 -

Here we instead start with a general type of disease, say infectious
disease and then try to find all GWAS studies that report on infectious
diseases. So, first we find the node/term in the EFO ontology that
corresponds to infectious disease. All infectious diseases should be
annotated in the EFO as descendants of that EFO term.

```{r tryChildren}
#first find the EFO term for infectious diseases
gg = grep("^infectious disease$", efo_df$Object, ignore.case=TRUE )
efo_df[gg,]

##now look at all the more specific terms
z2 = EFOdescendants("EFO:0005741")
length(z2)

efo_id = efo_df[z2,]
##d
## how many studies map directly to one of these terms
table(efo_id$Direct)
## one has 249 - and it is COVID, which makes some sense
efo_id[which.max(efo_id$Direct),]

##drop those with 0 mapping to them
efo_ids = efo_id[efo_id$Direct>0,]
##now how do we get all of those studies and do a Manhattan plot on them.

##now here I think one can just do a manhattan plot - maybe coloring covid differently?
dirMap = resources_annotated_with_term(efo_ids$Subject, include_subclasses=FALSE)
dim(dirMap)
```

## FIXME: need some way to look at the text descriptions and see if they are sensible

Where this can go wrong is that the synonyms might be inaccurate and
some investigation suggests that this true in many cases.

The next part of the process is to identify which, if any, of those
terms have studies in the GWAS catalog that actually map to that term.
The mapping can come in two ways: - direct: the study was mapped to that
term specifically - indirect: the study maps to a term because it mapped
directly to a more specific term and hence it is inferred to map to the
current ontology term.

Let's try to understand how we got to seven thousand studies and why 40
are mapped directly to Gout.

```{r exploreMappings}
##get all traits that mapped directly to CKD
z2 =getMatchedTraits(c("EFO:0003884"))

##we see gout show up in our datatable - 40 direct and 1 inherited
##this code gets the traits that map directly to Gout
dirMap = resources_annotated_with_term("EFO:0004274", include_subclasses=FALSE)
##first should be 40
length(unique(dirMap$STUDY.ACCESSION))
##second is 22 - 12/31/23 - 
length(unique(dirMap$PUBMEDID))
```

## Mapping to the Ontologies and finding the roll-ups

Here we look for GWAS hits that map to the EFO term `EFO:0009909`, which
is the trait *stage 5chronic kidney disease*. We use the
`resources_annotated_with_term` function to find all studies that are
mapped to this term.

```{r pressure, echo=FALSE}
ematch = resources_annotated_with_term("EFO:0009909", include_subclasses=FALSE)
dim(ematch)
allmatch = resources_annotated_with_term("EFO:0009909", include_subclasses = TRUE)
dim(allmatch)
DT::datatable(head(allmatch,n=20))
```

Now we can use the `variants_from_study` function to obtain information
about the GWAS hits such as the `rsID`, risk allele and effect size.

## Perhaps there are other terms that are interesting

```{r otherterms}
q2 = "Infectious AND disease"
hts2 = search_features(efo_tc, query = q2)
hitsasDT2 = hits2DT(hts2, efo_df, efo_tc)
DT::datatable(hitsasDT2[1:100,], escape=FALSE, rownames=FALSE)
```

Now lets get all the EFO terms and then get the resources that are
annotated at those terms. When obtaining the hits, we want to include
all the subclasses, so we get everything that is either mapped to
directly or a more specific description and hence should also be an
infectious disease. We will then drop any duplicated studies and extract
the SNPs reported for each study.

```{r getresources}
efo2 = unique(as.character(hts2$hits$doc_id))
ra2 = resources_annotated_with_term(efo2)

assoc2 <- granges_from_study(ra2$STUDY.ACCESSION)
assoc3 = assoc2[assoc2$P.VALUE<1e-8,]
##look at all the associations
NROW(assoc3)
##how many are reported more than once...
sum(duplicated(assoc3$SNP_ID_CURRENT))
##and produce a plot
pp = ggmanh::manhattan_data_preprocess(assoc3, pval.colname="P.VALUE")
ggmanh::manhattan_plot(pp, plot.title="Infectious Diseases")
```

Then one can also look at whether the same genes are implicated across
studies. Note that there are `r NROW(assoc3)` associations that are
plotted. The interested reaader can explore a number of different
aspects of these mappings.

```{r lookatHits}
sort(table(assoc3$MAPPED_GENE),dec=TRUE)[1:10]
tt = sort(table(assoc3$MAPPED_TRAIT_CURIE), dec=TRUE)
tt[1:10]
```

Then look at some of the traits that show up often. The top two hits are
for lymphocyte counts. Out of the top 10, only one seems to actually be
an infectious disease. The others are autoimmune, COPD, and measurements
of lymphocytes.

```{r maptotraits}
efo_df[names(tt)[1],]
efo_df[names(tt)[2],]

```

So, we might want to be able to find out how COPD gets mapped in to
infectious disease. Could these be the result of some of the more
complex comparisons that get reported. Maybe we need to cull the studies
that have multiple traits associated with them.

Returning to the example from the Quick Start vignette. Where we saw
that some studies have been mapped to more than one EFO term. For study
GCST008721, the title of the accompanying paper is *Genetic overlap
between autoimmune diseases and non-Hodgkin lymphoma subtypes*, which
suggests why this particular study is associated with two different EFO
terms.

So, we might want to include this study in an infectious disease
analysis, or in a study of the genetics of non-Hodgkin lymphoma, but we
probably would not want to pull in all papers that reported on
non-Hodgkin lymphoma.

Maybe what will work here is splitting by commas then trying to find out
if each term on its own will map to an infectious disease.

```{r multipletraits}
mtraits = assoc3$MAPPED_TRAIT_CURIE
g1 = grepl(",", mtraits, fixed=TRUE)
mtraits = unique(mtraits[g1])
```

## Pivot to a corpus based on the GWAS Catalog study IDs.

Here, we build a corpus based on the GWAS study IDs. The internal
function to do this is `.makeGWCcorpus`, which will create a corpus
where for each study (e.g. GCST006286), we that there are two entries
for this study representing two distinct mappings of the study disease
trait to different ontology terms.

```{r GWCexp1}
dbGetQuery = DBI::dbGetQuery
gwc_df <- dbGetQuery(gwasCatSearch_dbconn(), "SELECT * from gwascatalog_mappings")
gwc_df[which(gwc_df$STUDY.ACCESSION=="GCST006286"),]
```

We assemble the DISEASE.TRAIT, the MAPPED_TRAIT and all synonyms for
these terms that have been provided by the ontologies. These three sets
of descriptions are used when searching the corpus for matches.

Thus, we can find studies that have good matches to the input terms. But
this would not allow us to find studies that have been mapped to terms
that are related to these through the ontology. Here we are just making
use of the ontologies to give us synonyms.

```{r GWCexp2}
##find all the studies that map to leukemia
htl = search_features(gwc_tc, query="leuk*")
htlf = addField2Hits(htl, gwc_tc)
tt = table(htlf$hits$field)
tt
```

There were `r length(unique(htlf$hits$doc_id))` unique documents that
were identified. We can see that there were `r tt["DISEASE.TRAIT"]`
cases where a match was to the DISEASE.TRAIT, `r tt["Synonyms"]` matches
to synonyms provided by the ontologies, and `r tt["Trait"]` matches to
the text description used by the ontology for the trait.

## Can we search on Uberon categories.

Seems to be possible. Big question is how to get any sort of document
processing going -

```{r uberon}
lungU = "UBERON:0002048"
lungEFO = grepl(lungU, efo_df$DiseaseLocation)


efo_tc$meta$Lung = lungEFO

efo_lung = efo_df[lungEFO,]
##FIXME - here it would be good to get some form of graph completion of the EFO labels


efo_tcLung = subset(efo_tc, subset_meta = Lung == TRUE)
efo_tcLung2 = subset(efo_tcLung, subset=freq_filter(token, min=5))
dtmLung = get_dtm(efo_tcLung, "feature")
csums = apply(dtmLung, 2, sum, na.rm=T)
sort(csums/sum(csums),dec=T)[1:10]
dtm_wordcloud(dtmLung)
```

## Citations - to be fixed up

Nature . 2021 Nov;599(7886):628-634. doi: 10.1038/s41586-021-04103-z.
Epub 2021 Oct 18. Exome sequencing and analysis of 454,787 UK Biobank
participants PMID: 34662886
