---
title: "Ontology and Corpus Aided Searching of Complex Data Resources"
author: "RoGe, VC, RaGo"
date: "2023-06-23"
output: 
  html_document:
vignette: >
  %\VignetteIndexEntry{Ontology and Corpus Aided Searching of Complex Data Resources}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

bibliography: references.bib

link-citations: true

format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("gwasCatSearch")
library("DT")
```

## Introduction


Larger and more complex biomedical data sets are becoming increasingly
available and they provide substantial opportunities for discovery.
Realizing the potential for discovery is challenged by the complexity of
the data, the inherent relationships between phenotypes and the need for
data analysts to be able to navigate some of the complexities of both
biological and medical terminologies. In this paper we explore using
corpus search tools (as exemplified in the `corpustools` package
@atteveldtCorpustoolsManagingQuerying2023) to help search through the set of traits that are
reported by the GWAS catalog. These traits have been hand curated by the
EBI and mapped to terms in the Experimental Factor Ontology (EFO,
@maloneModelingSampleVariables2010). However, the tools available for searching these data are
mainly variants of regular expression matching, which as very limited
expressivity for the complex nature of medical and biological terms.

The GWAS Catalog (@sollisNHGRIEBIGWASCatalog2023) reports on a set of curated GWAS
association studies. Since 2010, delivery and development of the Catalog
has been a collaborative project between the EMBL-EBI and NHGRI. As of
2024-01-30, the GWAS Catalog contains 6715 publications, 571148 top
associations and 67322 full summary statistics. GWAS Catalog data is
currently mapped to Genome Assembly GRCh38.p14 and dbSNP Build 156.

Among the curation activities is a process of mapping the traits that
are studied to the EFO (@maloneModelingSampleVariables2010). In part this allows one to search for
studies about a particular disease, but for a variety of reasons this
search can be quite complicated. Here we report on tools and
approaches that can be used to help find GWAS studies according to
specific criteria. We create two searchable corpora, one that is EFO
centric and the other that is GWAS trait centric. There are a number of 
anomalies in both the EFO and the GWAS catalog that make it essential that
we have effective tools for cross-referencing searches. Our software is contained in the `gwasCatSearch` package (@gentlemanGwasCatSearchPackageSupport2024a) 
which is open source and freely available.

## Methods

### Data
We obtain data from the GWAS Catalog and transform it into a set of tables using tools in the GWASCatalogSearchDB (https://github.com/ccb-hms/GWASCatalogSearchDB). This processing produces
a number of tables that can be accessed via a SQL database; we use SQLite (@sqlite2020hipp). These tables include the raw data from the EBI, several tables organized around the Experimental Factor Ontology (EFO), the Uber-anatomy ontology (UBERON, https://www.ebi.ac.uk/ols4/ontologies/uberon).
An ontology is a set of concepts and categories that describe a subject area or domain of knowledge that expresses their properties and the relations between them. Ontologies form one of the bases for knowledge engineering and representation and are important in understanding and referencing biomedical entities. One of the challenges that needs to be addressed is the fact that the same concept might usefully appear in different ontologies and users will need tools to help identify similar concepts. A second challenge is that to be useful the ontologies must be accessible and typically that means using the web or a database implementation. There are two useful constructs that help with this, the compact uniform resource identifiers (CURIEs), and the internationalized resource identifiers (IRIs)  which in ontology space are used as *names* terms in the ontology. A CURIE has three parts: a prefix; a delimiter; and a unique local (to the ontology described by the prefix) identifier from the given nomenclature. The CURIE EFO:0005305 which represents the concept *atrioventricular node disease* has a prefix of EFO, a delimiter of a colon (:) and a unique local identifier of 0005305.  The corresponding IRI for that term is http://www.ebi.ac.uk/efo/EFO_0005305, which gives a web location that can be examined by a user.

FIXME: Rafael's description of how UBERON mapping is done - note one could try to build an UBERON based corpus...

As part of the processing we construct two dataframes, one based on the EFO terms `efo_df` and the other based on the GWAS catalog studies `gwc_df`. These are described in Table 1. We create two different corpora for searching. One based on the EFO and
the relationships inherent in it, `efo_tc`, and the second based on the GWAS
catalog studies, `gwc_tc`.

The EFO dataframe has one row for each term in the EFO ontology and columns labeled `Subject`: the corresponding EFO CURIE; `Object`: the textual description of the trait; `IRI`: the IRI for that term; `DiseaseLocation`: the UBERON CURIE for the location of the `trait` if there is one; `Direct`: the number of GWAS catalog studies that were mapped directly to that trait; `Inherited`: the number of traits that were mapped to terms that are more specific (according to the ontology) than the `Object`.

Since a GWAS study can be directly mapped to more than one EFO term we have made our
dataframe have one row for each unique combination of GWAS catalog study ID and EFO term.
The GWAS catalog dataframe has one row for each GWAS catalog study and columns labeled `STUDY.ACCESSION`: the GWAS catalog study identifier; `DISEASE.TRAIT`: the GWAS catalog text description of the trait; `MAPPED_TRAIT`: the textual description of the trait from EFO that this study was *directly* mapped to; `MAPPED_TRAIT_URI`: the URI for the mapped trait; `MAPPED_TRAIT_CURIE`: the CURIE for the mapped trait.

Each GWAS study has a specific trait that was studied and that is
reported as a text string. That trait is mapped to a specific term in
the EFO ontology, and we refer to those as **direct** mappings. The terms in the EFO ontology
are related to each other via a directed acyclic graph. By construction there is a single root node. More specific terms, such as the relationship between cancer and pancreatic cancer are encoded with pancreatic cancer being the **child** of cancer,
while cancer is the **parent** of pancreatic cancer.  The descriptions can become more refined, and we define the **descendents** of a node to be the set of child nodes, where one recurses down to the point where a node has no children.  Such a node is referred to as a *leaf*.  Similarly **ancesters** are the parents, again applied recursively to the root node of the ontology.  Our software contains tools to obtain the children, descendents, parents, and ancestors as well as to perform other operations on the ontology. 

Tools for representing and organizing biomedical entities and concepts are key components of
any system that process disease, medical or epidemiological data.  There are many challenges among them is the need to have some way to have two different systems align on some concepts. On the seminatic web Uniform Resource Identifiers (URIs) are used to reference named entities. So URI basically give a web address where more details on the concept can be found.

Given the structure of the ontology we can introduce the concept of an *inherited* mapping. A mapping at a node E is inherited if it was a direct mapping to some descendant of E.  Basically this is
providing a way to find, and hence examine, the EFO terms that map to more specific versions of an EFO term.  So, while we might want to know the GWAS catalog traits that map to cancer, we may also want to find the traits that map to more specific types of cancer.  This concept allows us to do that in a rigorous way.

### A search corpus based on EFO

To create a searchable EFO-based corpus we think of
each EFO term as a *document*. The EFO provides a text description of
the term as well as a set of known synonyms. We expand that by including the GWAS trait 
for any GWAS study that has been mapped **directly** to that EFO term.  Our rationale
is that the curators assigned that trait description to the specific EFO term, so it 
is essentially, in their eyes, a synonym.

For clarity, there is one document for each EFO term. The term itself we call the *object*,
that is augmented by a set of zero or more *synonyms* provided by EFO and then additionally
augmented by a set of zero or more *matches* which are GWAS study traits.  Below
we examine one node, EFO:0003086, which is labeled as chronic kidney disease.

```{r exampleEFO}
##these functions return named lists, one element for each provided EFO term
ckdSyns = getSynonyms("EFO:0003086")
ckdSyns
ckdMatches = getMatchedTraits("EFO:0003086")
length(ckdMatches[[1]])
ckdStudies = gwc_df[gwc_df$MAPPED_TRAIT_CURIE=="EFO:0003086",]
ckdStudies$STUDY.ACCESSION

```

At this point all we have done is asked which EFO terms match our
search. And by match our search we mean that one or more of the elements
of the text search is found in either the name of the EFO term, the EFO
provided synonyms for that term, or the names of the set of GWAS catalog
studies that the EBI has annotated with that same term.

## A search corpus based on GWAS Traits.

A dataframe, `gwc_df` containing all GWAS studies is provided with the `gwasCatSearch` package.
For each study (e.g. GCST006286), we report STUDY.ACCESSION, DISEASE.TRAIT, MAPPED_TRAIT,       MAPPED_TRAIT_URI, and MAPPED_TRAIT_CURIE. We note that for some studies there are multiple
rows, each one representing a unique direct mapping of that DISEASE.TRAIT to a unique EFO term.

```{r GWCexp1}
gwc_df[which(gwc_df$STUDY.ACCESSION=="GCST006286"),]
```


In this case we created documents by assembling the GWAS trait (DISEASE.TRAIT), the EFO mapped trait (Triat) and all EFO provided synonyms (Synonyms). Using this dataframe we construct a searchable corpus using the internal function `.makeGWCcorpus`.  The corpus can then find matches based on either of these three potentially different text descriptions of each of the GWAS traits.
We note that in this case the ontology is only being used to provide a trait and its synonyms, the relationships in the ontology are not used.

```{r GWCexp2}
##find all the studies that map to leukemia
htl = search_features(gwc_tc, query="leuk*")
htlf = addField2Hits(htl, gwc_tc)
tt = table(htlf$hits$field)
tt
```

There were `r length(unique(htlf$hits$doc_id))` unique documents that
were identified. We can see that there were `r tt["DISEASE.TRAIT"]`
cases where a match was to the DISEASE.TRAIT, `r tt["Synonyms"]` matches
to synonyms provided by the ontologies, and `r tt["Trait"]` matches to
the text description used by the ontology for the trait.

## Searching For Diseases

In this section we describe three different approaches that can be taken to
find GWAS studies that are associated with a particular term.  We will look for GWAS
studies associated with *infectious disease*.  One reason for this choice is that it is quite a challenging problem if one attempts it without using an ontology.  The main reason for that is that there are many infectious diseases and most of them do not have either *infectious* or *disease* in their name.  So without an ontology to guide us, we would need to have some dictionary of infectious diseases to start with. [FIXME: should we try to find one?]

### Strategy One: Use the EFO

Here describe how to use these tools start with a general type of disease, say infectious
disease, and then try to find all GWAS studies that report on that class of 
diseases. So, first we find the node/term in the EFO ontology that
corresponds to infectious disease. All infectious diseases should be
annotated in the EFO as descendants of that EFO term.

```{r tryChildren}
#first find the EFO term for infectious diseases
gg = grep("^infectious disease$", efo_df$Object, ignore.case=TRUE )
efo_df[gg,]

##now look at all the more specific terms
z2 = EFOdescendants("EFO:0005741")
length(z2)

efo_id = efo_df[z2,]
##d
## how many studies map directly to one of these terms
table(efo_id$Direct)
## one has 249 - and it is COVID, which makes some sense
efo_id[which.max(efo_id$Direct),]

##drop those with 0 mapping to them
efo_ids = efo_id[efo_id$Direct>0,]

dirMap = resources_annotated_with_term(efo_ids$Subject, include_subclasses=FALSE)
dim(dirMap)
```

 FIXME: need some way to look at the text descriptions and see if they are sensible
 
### Strategy Two: Search based on the GWAS catalog corpus

For this search we start fairly broadly and search the `gwc_tc` for the term *infectious*. 
Recall that this will search the text descriptions of the GWAS trait as well as the trait and synonyms that correspond to the EFO term(s) this study was directly mapped to.

```{r IDgwctc}
qID = "infectious"
htsID = search_features(gwc_tc, query = qID)
htsID = addField2Hits(htsID, gwc_tc )
hitsIDasDT = hits2DTGWC(htsID, gwc_df, gwc_tc)
```

We can see where the matches occur. And see that the bulk have come from the EFO synonyms.

```{r checkIDmatches}
idTT = table(hitsIDasDT$field)
idTT
```


Now we can ask if this second approach found any GWAS studies that the first one did not.
And further look at those, to see if any are likely to be about infectious diseases. We
will also check to see if the first method found studies that the second approach did not.

```{r compID}
vv = setdiff(hitsIDasDT$Study, dirMap$STUDY.ACCESSION )
v2 = setdiff(dirMap$STUDY.ACCESSION, hitsIDasDT$Study)

DT::datatable(hitsIDasDT[match(vv, hitsIDasDT$Study),])
```

So there are `r length(vv)` studies that were found by the second method, that were not found by the first.  And `r length(v2)` studies found by the first, but not the second.  Which is probably what we hoped for. Many infectious diseases do not have the the words *infectious disease* as part of their name.

### Strategy three: A text search on the EFO ontology

Above we first started with the *infectious disease* in the EFO ontology and then found
all GWAS studies that mapped directly to either that node, or one of its descendants. We contrasted that with the strategy of searching the corpus based on the GWAS catalog studies for the terms infectious and disease.  A third approach is to search the EFO corpus for *infectious disease*
which we do in the code below

```{r otherterms}

qIDefo = "Infectious AND disease"
htsID2 = search_features(efo_tc, query = qIDefo)
hitsID2asDT2 = hits2DT(htsID2, efo_df, efo_tc)
##DT::datatable(hitsasDT2[1:100,], escape=FALSE, rownames=FALSE)

```

Now we can compare the EFO categories and see if we found anything new and also examine
the GWAS studies. A quick visual inspection of the EFO categories found by this method
and not the first, suggests that the first approach did seem to get everything useful.

```{r method3comp}
efoIDS3 = row.names(hitsID2asDT2)
setd3 = setdiff(efoIDS3, row.names(efo_id))
efo_df[setd3,"Object"]

```

We end this section by creating a Manhattan plot based on the studies that were identified using the first strategy. 

```{r getresources}

IDassoc <- granges_from_study(dirMap$STUDY.ACCESSION)
IDassoc2 = IDassoc[IDassoc$P.VALUE<1e-8,]

NROW(IDassoc2)
##how many are reported more than once...
sum(duplicated(IDassoc2$SNP_ID_CURRENT))
##and produce a plot
```

```{r IDmanh, plot=TRUE, fig.align='center', fig.width=7, fig.cap="Manhattan Plot for Infectious Disease", warning=FALSE, fig.height=5, fig.path="figure/"}
pp = variants_from_study(dirMap$STUDY.ACCESSION)
pp = pp[IDassoc$P.VALUE<1e-8,]

#pp = ggmanh::manhattan_data_preprocess(IDassoc2, pval.colname="P.VALUE")
simple_ggmanh(pp, y.label="-log10 p",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

##ggmanh::manhattan_plot(pp, plot.title="Infectious Diseases", y.label="-log10 p")
```

## Search for GWAS studies by Disease Location

In recent versions of the EFO the curators have begun to include disease location information.
This information is encoded using the UBERON ontology. To use this functionality you would first identify the UBERON CURIE that represents the location of interest.  In the code below we look for diseases associated with the lung.  FIXME: Vince - any sort of EFO graph that we could use to represent the disease categories associated with lung?

In the code below we identify the appropriate CURIE and then find all EFO terms that have been annotated with that term. FIXME - can there be sub-terms? and do we need some tools to roll up?
We next add a column to the metadata for the `efo_tc` corpus. This then allows us to subset the corpus and then we can carry out analyses based on the subset. Details on these capabilities are given in the @atteveldtCorpustoolsManagingQuerying2023 package vignette.

```{r uberon, warning=FALSE}
lungU = "UBERON:0002048"
lungEFO = grepl(lungU, efo_df$DiseaseLocation)


efo_tc$meta$Lung = lungEFO

efo_lung = efo_df[lungEFO,]
##FIXME - here it would be good to get some form of graph completion of the EFO labels

```

In the code below we subset first to the set of terms that have an association with the lung and then to the set of tokens that have more than 5 occurrances. And finally we create a word cloud based on the frequency of the tokens that remain.  These are based on occurance in the EFO corpus.

```{r subsetLung, warning=FALSE}
efo_tcLung = subset(efo_tc, subset_meta = Lung == TRUE)
efo_tcLung2 = subset(efo_tcLung, subset=freq_filter(token, min=5))
dtmLung = get_dtm(efo_tcLung, "feature")
csums = apply(dtmLung, 2, sum, na.rm=T)
sort(csums/sum(csums),dec=T)[1:10]
dtm_wordcloud(dtmLung)
```



## Exploring the GWAS metadata

Once we have a set of GWAS studies that we are interested in we can access the metadata
about which variants are implicated, where they are in the genome, the p-value and effect size to name a few.  This data is quite large and we use a SQLite database as the main tool for storing and accessing the data. The data are accessed via either `granges_from_study` if you want a `granges` object that can be integrated with a variety of Bioconductor @ tools for analyzing and plotting genomic data, or `variants_from_study` which returns a dataframe with 21 columns providing a set of metadata.

```{r from_studyfuncs}

tab = granges_from_study(acc="GCST007087")
tab3 = variants_from_study(acc="GCST007087")
names(tab3)
##do we care that these two are not the same dimenions?  

```

Then one can also look at whether the same genes are implicated across
studies. Note that there are #r NROW(assoc3)# associations that are
plotted. The interested reaader can explore a number of different
aspects of these mappings.

```{r lookatHits, eval=FALSE}
sort(table(assoc3$MAPPED_GENE),dec=TRUE)[1:10]
tt = sort(table(assoc3$MAPPED_TRAIT_CURIE), dec=TRUE)
tt[1:10]
```

Then look at some of the traits that show up often. The top two hits are
for lymphocyte counts. Out of the top 10, only one seems to actually be
an infectious disease. The others are autoimmune, COPD, and measurements
of lymphocytes.

```{r maptotraits, eval=FALSE}
efo_df[names(tt)[1],]
efo_df[names(tt)[2],]

```



## Challenges
1) the issue found with CKD
2) Vague study definitions
3) Inaccuracies in EFO and other metadata

## Vague Study Definitions

A second challenge that a data analyst is must face is that some of EFO terms that were used are too
vague. By that we mean that they have too many studies that map to them and hence are unlikely to be informative about the sort of specific questions we might want to address. These are
relatively easy to find. We used the cutoff of 500 studies, and in the code/table below we show the set of EFO terms that have more than 500 GWAS studies that have been directly mapped to them. 


```{r studies2EFO}

manystudies = efo_df[efo_df$Direct>500,]
DT::datatable(manystudies[,c(1,2,5,6)], rownames=FALSE)


```

One can then get a sense of the actual traits that were mapped to these by using one querying the GWAS catalog dataframe. We will look at EFO:0004725, which is *metabolite measurement*. In the code below we see that this EFO term is really a catch-all for a wide variety of chemical measurements and hence unlikely to be helpful providing insight into the commonality of terms that were mapped directly to it. 

##FIXME: Ask Rafael about other ideas here - maybe there is some sort of diversity index that could be computed per EFO term that says something about how related the terms mapped there are

```{r vagueTraits}
metabTraits = gwc_df[gwc_df$MAPPED_TRAIT_CURIE=="EFO:0004725", "DISEASE.TRAIT"]
zz = table(metabTraits)
sort(zz,dec=T)[1:10]

```

### Mapping not so great
In this example we are looking for kidney function traits, and in particular we
will search for *chronic kidney disease* in the EFO-based corpus.

```{r gwasCatPubMed}
q1 = "chronic AND kidney AND disease"
hts = search_features(efo_tc, query = q1)
hts = addField2Hits(hts, efo_tc )
hitsasDT = hits2DT(hts, efo_df, efo_tc)
```

#### Table 1. CKD EFO Hits (first 100)
```{r DT, fig.cap="CKD EFO Hits", echo=FALSE}
DT::datatable(hitsasDT[1:100,], escape=FALSE, rownames=FALSE)
```

After running the code above we find that there were `r nrow(hitsasDT)`
different EFO terms that have a hit. This seems very large. In the next
code chunk we explore some of the issues. We first look for a EFO term that
is just *kidney disease*, since we anticipate that all different form so kidney
disease will be descendants of that term (node in the EFO graph).

```{r ckdEFO}
kdEFO = grep("^kidney disease$", efo_df$Object, ignore.case = TRUE)
efo_df[kdEFO,]
ackd = EFOdescendants("EFO:0003086")

```

So, there are only `r length(ackd)` EFO terms that are more specific than the term *kidney disease* so it is a bit surprising that there are so many EFO terms that we found with our search.  It is also surprising that if we add the total number of GWAS studies that mapped directly to EFO:0003086 or were inherited by mapping to more specific terms we only get `r efo_df["EFO:0003086","Direct"] + efo_df["EFO:0003086","Inherited"]` GWAS studies.  This suggests that there are a very large number of studies where the GWAS trait contains the phrase *chronic kidney disease* that were not mapped to any EFO kidney disease. 

We next try to understand what is happening.  Recall that we constructed our documents by combining the name of the EFO trait (Object), the EFO synonyms (Synonyms) and all GWAS traits that were directly mapped to that EFO trait (Matches) into one *document*.  We can query the corpus to determine which of these three parts any match was made to.
the matches to *chronic kidney disease* are occurring. For example, for `r hitsasDT$EFO[1]`
the string *chronic kidney disease* was matched only to other GWAS
catalog traits (since only matches were found).

```{r why7K}
table(hitsasDT$field)
```

This means that the bulk of the EFO terms are showing up due to finding
the text string in the name of some other GWAS catalog trait and for some reason
that trait was not mapped to kidney disease. We next query the GWAS catalog traits.
The dataframe `gwc_df` contains the names of all traits in the `DISEASE.TRAIT` column.

```{r gwcmetadtable}
gg = grep("chronic kidney disease", gwc_df$DISEASE.TRAIT, ignore.case=T)

##gwc_df[gg[1:10], "DISEASE.TRAIT"]
ckdPMIDs = GWCATID2PMID(unique(gwc_df$STUDY.ACCESSION[gg]))
pmtab = sort(table(ckdPMIDs), dec=TRUE)
pmtab[1:10]
```

And now we see that there are in fact `r length(unique(gwc_df[gg, "STUDY.ACCESSION"]))` terms that have been labeled this way. Further, if we look at the PUBMEDIDs associated
with these traits we see that there are `r sum(pmtab>1000)` papers that
report more than 1000 traits. Given the PubMed IDs it is easy to find
the papers. The paper associated with PMID 35870639 which reports more than 
6,700 studies is titled: Identification of 969 protein quantitative trait loci in an African American population with kidney disease attributed to hypertension @@surapaneniIdentification969Protein2022.


We end this example by creating a Manhattan plot based on the studies that contain the string *chronic kidney disease*.

```{r ckdmanh, plot=TRUE, fig.align='center', fig.width=7, fig.cap="Manhattan Plot for CKD", warning=FALSE, fig.height=5, fig.path="figure/"}
vars = variants_from_study(gwc_df$STUDY.ACCESSION[gg])
##drop those with a p-value > 1e-8,
vars = vars[vars$`P-VALUE` < 1e-8, ]
simple_ggmanh(vars, y.label="-log10 p", plot.title = "Chronic Kidney Disease",
       label.colname = "MAPPED_TRAIT", pval.colname="P-VALUE")

```

### Bad synonyms

Another potential issue is that the EFO provided synonyms might be inaccurate. Further investigation of those issues is also recommended. FIXME: maybe Rafael has some ideas.


## Citations



