---
title: "Paper"
author: "RG"
date: "2023-06-23"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Paper}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
format: html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("gwasCatSearch")
library("DT")
```

## Example 1
Example in kidney function traits (Li et al., 2020) PMID 32764137

The query

(glomerul* AND filtration) OR urate OR (chronic AND kidney AND disease) OR (urea AND nitrogen) produces, via search_features and hits2DT, 7 ontologic tags:


```{r }
q1 = "(glomerul* AND filtration) OR urate OR (chronic AND kidney AND disease) OR (urea AND nitrogen)"
hts = search_features(efo_tc, query = q1)
hitsasDT = hits2DT(hts, efo_df, efo_tc)
DT::datatable(hitsasDT, escape=FALSE, rownames=FALSE)
```

At this point all we have done is asked which terms in the ontology match our search. The next part of the process is to identify which, if any, of those terms have studies in the GWAS catalog that actually map to that term. 
The mapping can come in two ways:
- direct: the study was mapped to that term specifically
- indirect: the study maps to a term because it mapped directly to a more specific term and hence it is inferred to map to the current ontology term.

## Mapping to the Ontologies and finding the roll-ups


```{r pressure, echo=FALSE}
ematch = resources_annotated_with_term("EFO:0009909", include_subclasses=FALSE)
dim(ematch)
allmatch = resources_annotated_with_term("EFO:0009909", include_subclasses = TRUE)
dim(allmatch)
DT::datatable(head(allmatch,n=20))
```

Now we can use the `variants_from_study` function to obtain information about the GWAS hits such as the `rsID`, risk allele and effect size.

```{r lookupSNP}
lu1 = variants_from_study(allmatch$STUDY.ACCESSION)
names(lu1)
lu1$MAPPED_GENE
```

And then we can visualize these variants using `simple_ggmanh`:

```{r plotContext, echo=FALSE, eval=TRUE}
simple_ggmanh(lu1, label.colname="MAPPED_TRAIT")
```

## Perhaps there are other terms that are interesting

```{r otherterms}
q2 = "Infectious AND disease"
hts2 = search_features(efo_tc, query = q2)
hitsasDT2 = hits2DT(hts2, efo_df, efo_tc)
DT::datatable(hitsasDT2, escape=FALSE, rownames=FALSE)
```

