---
title: "Working with Ontologies"
author: "Robert"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Working with Ontologies}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
format: html
editor: visual
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(gwasCatSearch)
```

## Using the ontologies

We have functions that allow exploration of the EFO ontology along the lines of the ontologyIndex package from CRAN. In the code below we start with an ontology term print its name and then find the parents and children of that term.

The ontology index contains information about children, parents and ancestors for different terms in an ontology. By default the tool uses *is-a* relationships to define these terms. The ontology_index object is just a list of 'vectors and lists' of term properties, indexed by the IDs of the terms:

```{r holder, echo=FALSE}
##    property     class
## 1        id character
## 2      name character
## 3   parents      list
## 4  children      list
## 5 ancestors      list
## 6  obsolete   logical
```

```{r , echo=TRUE}
EFOlabels("EFO:0000095")
EFOchildren("EFO:0000095")
EFOparents("EFO:0000095")
```

We'll produce a "little" ontology_index instance for illustration.

```{r dolit}
litoi = make_oi("EFO:0000095") 
litoi
```

One of the parents of `EFO:0000095` is `EFO:0000220` and we can see how many children it has by using the `children` table in the ontology index.

```{r chk}
litoi$children["EFO:0000095"]
```

If we use the `[[` accessor then we just get the vector of children. If instead we use the `[` accessor then we would get a list that is named by the input vector of terms where each element is the vector of children for the corresponding EFO term.

```{r}
x1 = litoi$children[["EFO:0000220"]]

x1
```

We see that "EFO:0000095" is one of the children of "EFO:0000220", as we expect given the fact that "EFO:0000220" is a parent of "EFO:0000095"

Now we will interact with the DB to get at the mappings to different terms. In the code below we can find which of the children of "EFO:0000220" have studies that are directly mapped to them or that have inherited a mapping from. one of its children.

```{r dbmappings}

 query = paste0("SELECT * FROM efo_labels WHERE Subject IN ('",
                paste(x1, collapse = "','"),
                "')")
 mappedans = DBI::dbGetQuery(gwasCatSearch_dbconn(), query)
 mappedans$Direct
 mappedans$Inherited

```

In the code chunk below we get the counts for the parents of "EFO:0000095". Suppose we are interested in this phenotype and interpreting the GWAS hits. One approach we could take to interpreting any GWAS hits for this phenotype is to download GWAS summary statistics for this phenotype (if it was computed) from the UKBB. We could then do either a Manhattan plot of those summary statistics in a local neighborhood, or we could do a compressed view that covered all chromosomes.

We might also wonder if there are GWAS hits for related phenotypes that are in similar locations in the genome. For example, since this phenotype is "chronic lymphocytic leukemia" we might ask if we look at all the hits mapped to the leukemia node in our ontology would they show any shared location.

Our goal is to find a set of GWAS phenotypes that

```{r parentmappings, eval=FALSE}
query = paste0("SELECT * FROM efo_labels WHERE Subject IN ('",
                paste(litoi$parents[["EFO:0000095"]], collapse = "','"),
                "')")
 pans = DBI::dbGetQuery(gwasCatSearch_dbconn(), query)
 pans$Direct
 pans$Inherited
```

Now we can look at some plotting for the graph and manipulations of the EFO tables.

```{r graphPlot, eval=FALSE}
#tags = c("EFO:0003884", "EFO:0004531", "EFO:0005208", "EFO:0009909", "EFO:0021409", #"EFO:0021433")
##assign the next one to get back the graph structure that is plotted
#get "EFO:0000095" and
tags = c(x1, "EFO:0000095")
##plotTags seems to cause some big download
xt = plotTags(tags) ## could use make_oi and onto_plot2
nn = graph::nodes(xt)
##nodes have long names - suitable for printing we just want the CURIEs
v1 = strsplit(nn, "\n")
nodeEFO = sapply(v1, function(x) x[length(x)])
subs=efo_df[nodeEFO,c("Direct","Inherited")]
subs
```

Some semantic network code

```{r semnet, warning=FALSE, message=FALSE, echo=FALSE}
gwc_tc\$preprocess(min_docfreq = 50, remove_stopwords = T, remove_numbers = T)

g = semnet_window(gwc_tc, 'feature')
gb = backbone_filter(g, alpha = 0.001, max_vertices = 100)
plot_semnet(gb)

fa = feature_associations(gwc_tc, 'feature', query = 'chron\*')
fa = feature_associations(gwc_tc, 'feature', query = 'chronic AND kidney')

fa = feature_associations(gwc_tc, 'feature', query = "leuk\*")
```
